### PR TITLE
Various fetchcargo improvements

### PR DESCRIPTION
**Pull request by alyssais**
###### Motivation for this change

- Most stdenv wrappers already forward unused arguments -- it allows greater customisation.  We just have to be careful to remove arguments we're using that shouldn't be passed to stdenv.  I've been conservative here, because fetchcargo checksums shouldn't change lightly.

- If a custom unpackPhase is used for the package, it needs to also be used for fetchcargo so the same source is available for vendoring.

- I exposed `fetchcargo` as `rustPlatform.fetchcargo`, because it might be useful to provide from another derivation.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

### CODE REVIEW

File: pkgs/build-support/rust/default.nix [hunk 1]
```diff
@@ -29,7 +30,7 @@ assert buildType == "release" || buildType == "debug";
 let
   cargoDeps = if cargoVendorDir == null
     then fetchcargo {
-        inherit name src srcs sourceRoot cargoUpdateHook;
+        inherit name src srcs sourceRoot unpackPhase cargoUpdateHook;
```
Comment (Mic92) [inline pkgs/build-support/rust/default.nix]: It's just a bit counter-intuitive that `unpackPhase` is executed on the `-vendor` directory and not on the package itself. Should it be called on both?

Comment (alyssais) [reply pkgs/build-support/rust/default.nix]: 2c318ca1b42b0b806ba5a4e328d193a0fba9581f makes it execute on the package itself as well.

Comment (alyssais) [reply pkgs/build-support/rust/default.nix]: It’s already called on the package itself because args to `buildRustPackage` are passed through to `stdenv`.

**Comment by Mic92** [comment]
Can you provide an usecase example, how you would use this?

**Comment by alyssais** [comment]
> Can you provide an usecase example, how you would use this?

Sure.

Here I use a custom unpackPhase for a Rust program that is very fussy
about where it is on disk in relation to its dependencies.  Without the
changes in this PR, doing this isn't possible because fetchcargo doesn't
know about the custom unpackPhase.

https://spectrum-os.org/git/nixpkgs/tree/pkgs/os-specific/linux/chromium-os/crosvm/default.nix

We have an open PR to add that package to Nixpkgs (#52353), and as you
can see it has to do something much more complicated, using several
intermediary src derivations, to work around this limitation of
fetchcargo.

File: pkgs/development/compilers/rust/default.nix [hunk 2]
```diff
@@ -19,17 +19,17 @@
...
-      inherit rustc cargo;
+    fetchcargo = buildPackages.callPackage ../../../build-support/rust/fetchcargo.nix {
+      inherit cargo;
+    };
 
-      fetchcargo = buildPackages.callPackage ../../../build-support/rust/fetchcargo.nix {
-        inherit cargo;
-      };
+    buildRustPackage = callPackage ../../../build-support/rust {
+      inherit rustc cargo fetchcargo;
```
Comment (7c6f434c) [inline pkgs/development/compilers/rust/default.nix]: How this passed `fetchcargo` get used?

Comment (alyssais) [reply pkgs/development/compilers/rust/default.nix]: Not sure I understand? It’s used here:

https://github.com/NixOS/nixpkgs/blob/83ac9c07e4df352e177ebdf978320089a137183a/pkgs/build-support/rust/default.nix#L31

Comment (7c6f434c) [reply pkgs/development/compilers/rust/default.nix]: Oh my bad, I looked at the changes, that you add `fetchcargo` argument but do not add any uses of it. Now I see that the point is to pass the local version referencing correct `cargo`.

**Review by 7c6f434c — APPROVED** [review]

**Pull request merged**